modulus/datapipes/climate/climate_hdf5.py

```diff
@@ -0,0 +1,591 @@
+        Shuffle dataset, by default True
+    num_workers : int, optional
+        Number of workers, by default 1
+    device: Union[str, torch.device], optional
+        Device for DALI pipeline to run on, by default cuda
+    process_rank : int, optional
+        Rank ID of local process, by default 0
+    world_size : int, optional
+        Number of training processes, by default 1
+    """
+
+    def __init__(
+        self,
+        data_dir: str,
+        stats_dir: Union[str, None] = None,
+        channels: Union[List[int], None] = None,
+        batch_size: int = 1,
+        stride: int = 1,
+        dt: float = 6,
+        start_year: int = 1980,
+        num_steps: int = 2,
+        lsm_filename: str = None,
+        geopotential_filename: str = None,
+        use_latlon: bool = False,
+        use_cos_zenith: bool = False,
+        patch_size: Union[Tuple[int, int], int, None] = None,
+        num_samples_per_year: Union[int, None] = None,
+        shuffle: bool = True,
+        num_workers: int = 1,
+        device: Union[str, torch.device] = "cuda",
+        process_rank: int = 0,
+        world_size: int = 1,
+    ):
+        super().__init__(meta=MetaData())
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.shuffle = shuffle
+        self.data_dir = Path(data_dir)
+        self.stats_dir = Path(stats_dir) if not stats_dir is None else None
+        self.channels = channels
+        self.stride = stride
+        self.dt = dt
+        self.start_year = start_year
+        self.num_steps = num_steps
+        self.lsm_filename = lsm_filename
+        self.geopotential_filename = geopotential_filename
+        if use_cos_zenith:
+            use_latlon = True
+        self.use_latlon = use_latlon
+        self.use_cos_zenith = use_cos_zenith
+        self.process_rank = process_rank
+        self.world_size = world_size
+        if isinstance(patch_size, int):
+            patch_size = (patch_size, patch_size)
+        self.patch_size = patch_size
+        self.num_samples_per_year = num_samples_per_year
+
+        # Determine outputs of pipeline
+        self.pipe_outputs = ["state_seq", "timestamps"]
+        if self.lsm_filename is not None:
+            self.pipe_outputs.append("land_sea_mask")
+        if self.geopotential_filename is not None:
+            self.pipe_outputs.append("geopotential")
+        if self.use_latlon:
+            self.pipe_outputs.append("latlon")
+            self.pipe_outputs.append("cos_latlon")
+        if self.use_cos_zenith:
+            self.pipe_outputs.append("cos_zenith")
+
+        # Set up device, needed for pipeline
+        if isinstance(device, str):
+            device = torch.device(device)
+
+        # Need a index id if cuda
+        if device.type == "cuda" and device.index == None:
+            device = torch.device("cuda:0")
+        self.device = device
+
+        # check root directory exists
+        if not self.data_dir.is_dir():
```

Comment:
What if the data_dir is not on some local file system? Should we use something like fsspec for generality?

Reply:
This is a good point and we have some file abstraction already in Modulus. Right now though this datapipe should definitely not be used for any non-local file systems. If you tried this with s3 for example it would be very slow. There are some other issue that come up with pulling data with multiple threads as well that might cause issues.